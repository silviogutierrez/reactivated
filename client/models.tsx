import {WidgetType} from "./components/Widget";
import {FieldType, FormViewProps} from "../exports";

declare module "../exports" {
    interface FieldType {
        widget: WidgetType;
    }
}

export * from "../exports";
