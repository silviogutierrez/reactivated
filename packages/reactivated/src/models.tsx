export interface Settings {
    DEBUG_PORT: number;
    MEDIA_URL: string;
    STATIC_URL: string;
    DEBUG: boolean;
}

export interface Message {
    level: number;
    level_tag: "info" | "success" | "error" | "warning" | "debug";
    message: string;
}
