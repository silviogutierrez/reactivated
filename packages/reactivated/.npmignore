dist
*.ts
*.tsx
!*.d.ts
