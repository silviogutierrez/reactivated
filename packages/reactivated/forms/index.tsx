import React from "react";

export {FormLike, FormSetLike, FieldMap, ManagementForm} from "../components/Form";
export {Field} from "../components/Field";

import {Field} from "../components/Field";
import {FieldMap, FormLike} from "../components/Form";
import {SelectDateWidget, Widget, WidgetType} from "../components/Widget";
import Context from "../context";

export const CSRFToken = (props: {}) => {
    const context = React.useContext(Context);

    return (
        <input type="hidden" name="csrfmiddlewaretoken" value={context.csrf_token} />
    );
};

type FormValue<U extends WidgetType> = U extends SelectDateWidget
    ? {[subWidgetName: string]: U["subwidgets"][number]["value"]}
    : U["value"];

export type FormValues<U extends FieldMap> = {
    [K in keyof U]: FormValue<U[K]["widget"]>;
};

interface BaseFieldsProps<U extends FieldMap> {
    fieldInterceptor?: (form: FormLike<U>, fieldName: keyof U) => U[keyof U];
    changeInterceptor?: (
        name: keyof U,
        prevValues: FormValues<U>,
        nextValues: FormValues<U>,
    ) => FormValues<U>;
    form: FormLike<U>;
    children: (props: {error: string[] | null; field: U[keyof U]}) => React.ReactNode;
}

interface IncludeFieldsProps<U extends FieldMap> extends BaseFieldsProps<U> {
    fields?: Array<keyof U>;
    exclude?: never;
}

interface ExcludeFieldProps<U extends FieldMap> extends BaseFieldsProps<U> {
    fields?: never;
    exclude: Array<keyof U>;
}

type FieldsProps<U extends FieldMap> = IncludeFieldsProps<U> | ExcludeFieldProps<U>;

export const Fields = <U extends FieldMap>(props: FieldsProps<U>) => {
    const getField =
        props.fieldInterceptor ?? ((form, fieldName) => form.fields[fieldName]);

    const getIterator = () => {
        if (props.fields != null) {
            return props.fields;
        }

        if (props.exclude != null) {
            return props.form.iterator.filter(
                (field) => !props.exclude.includes(field),
            );
        }

        return props.form.iterator;
    };

    const iterator = getIterator();

    return (
        <>
            {iterator
                .map(
                    (fieldName) =>
                        [fieldName, getField(props.form, fieldName)] as const,
                )
                .map(([fieldName, field]) => {
                    const {widget} = field;
                    const error =
                        props.form.errors != null
                            ? props.form.errors[fieldName] ?? null
                            : null;

                    return (
                        <React.Fragment key={fieldName.toString()}>
                            {props.children({field, error})}
                        </React.Fragment>
                    );
                })}
        </>
    );
};
