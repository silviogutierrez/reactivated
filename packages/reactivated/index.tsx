export * from "./components/Form";
export {FormSet} from "./components/FormSet";
export * from "./components/SectionalForm";
export * from "./components/Widget";
