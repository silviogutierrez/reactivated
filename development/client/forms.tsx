import React from "react";

import {css} from "@linaria/core";
import {styled} from "@linaria/react";

import {FieldHandler, Widget} from "reactivated/forms";

import {Types} from "@client/generated";
import * as styles from "@client/styles";

export {useFormSet, ManagementForm} from "reactivated/forms";

export const Field = (props: {field: FieldHandler<Types["globals"]["Widget"]>}) => {
    const {field} = props;
    const renderedWidget = <Widget field={field} />;

    if (field.tag == "django.forms.widgets.HiddenInput") {
        return renderedWidget;
    }

    return (
        <label
            className={css`
                ${styles.verticallySpaced(5)}
                display: block;
            `}
        >
            <div
                className={css`
                    font-weight: 700;
                `}
            >
                {field.label}
            </div>
            {renderedWidget}
            {field.error != null && (
                <div
                    className={css`
                        color: #cf0000;
                    `}
                >
                    {field.error}
                </div>
            )}
        </label>
    );
};

export const Fieldset = styled.fieldset`
    border: 1px solid #bbb;
    border-radius: 5px;
    padding: 20px;
`;

export const Button = styled.button`
    border: 1px solid #bbb;
    border-radius: 5px;
    padding: 10px 15px;
    font: inherit;
    text-transform: lowercase;
    font-weight: 700;
    background-color: white;
    color: #444;
    cursor: pointer;
`;

export const ButtonLink = styled.a`
    display: inline-block;
    border: 1px solid #bbb;
    border-radius: 5px;
    padding: 10px 15px;
    font: inherit;
    text-transform: lowercase;
    font-weight: 700;
    background-color: white;
    color: #444;
    cursor: pointer;
    text-decoration: none;
`;
