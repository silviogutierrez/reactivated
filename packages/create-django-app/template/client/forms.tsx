import React from "react";

import {css} from "@linaria/core";

import {FieldHandler, Widget} from "reactivated/forms";

import {Types} from "@client/generated";

export const Field = (props: {field: FieldHandler<Types["globals"]["Widget"]>}) => {
    const {field} = props;
    const renderedWidget = <Widget field={field} />;

    if (field.tag == "django.forms.widgets.HiddenInput") {
        return renderedWidget;
    }

    return (
        <label
            className={css`
                display: block;
            `}
        >
            <div>{field.label}</div>
            {renderedWidget}
            {field.error != null && (
                <div
                    className={css`
                        color: red;
                    `}
                >
                    {field.error}
                </div>
            )}
        </label>
    );
};
