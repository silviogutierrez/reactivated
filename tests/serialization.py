import enum
from typing import Any, List, Literal, NamedTuple, Tuple, Type

import pytest
import simplejson
from django import forms as django_forms
from django.apps.registry import Apps
from django.db.models import IntegerField, Model
from django.forms.models import ModelChoiceIteratorValue
from jsonschema import validate

from reactivated import Pick
from reactivated.serialization import create_schema, serialize
from sample.server.apps.samples import forms, models


class Spam(NamedTuple):
    thing: List[str]
    again: str


class Bar(NamedTuple):
    a: str
    b: bool


class SlimEnum(enum.Enum):
    FIRST = "Ok"
    SECOND = "Great"


class ChunkyEnumMember(NamedTuple):
    is_important: bool
    title: str


class ChunkyEnum(enum.Enum):
    CHUNKY_FIRST = ChunkyEnumMember(is_important=False, title="Chunky First")
    CHUNKY_SECOND = ChunkyEnumMember(is_important=True, title="Chunky Second")


Opera = Pick[models.Opera, "name"]


class Foo(NamedTuple):
    bar: Bar
    spam: Spam

    slim_enum_type: Type[SlimEnum]
    chunky_enum_type: Type[ChunkyEnum]

    pick: Pick[models.Composer, "name", "operas.name"]
    pick_many: List[Pick[models.Composer, "name", "operas.name"]]
    pick_method: Pick[models.Opera, "name", "get_birthplace_of_composer"]
    pick_property: Pick[models.Composer, "did_live_in_more_than_one_country"]
    pick_literal: Pick[models.Composer, Literal["name", "operas.name"]]
    pick_computed_queryset: Pick[
        models.Composer, "operas_with_piano_transcriptions.name"
    ]
    pick_nested: Pick[models.Composer, "name", Pick["operas", Opera]]
    pick_enum: Pick[models.Continent, "hemisphere"]
    fixed_tuple_different_types: Tuple[str, int]
    fixed_tuple_same_types: Tuple[str, str]
    complex_type_we_do_not_type: Any


def convert_to_json_and_validate(instance, schema):
    converted = simplejson.loads(simplejson.dumps(instance))
    validate(
        instance=converted, schema={"definitions": schema.definitions, **schema.schema}
    )


@pytest.mark.django_db
def test_serialization():
    continent = models.Continent.objects.create(name="Europe")
    birth_country = models.Country.objects.create(name="Germany", continent=continent)
    other = models.Country.objects.create(name="Switzerland", continent=continent)

    composer = models.Composer.objects.create(name="Wagner")
    models.ComposerCountry.objects.create(
        composer=composer, country=birth_country, was_born=True
    )
    models.ComposerCountry.objects.create(composer=composer, country=other)

    opera = models.Opera.objects.create(name="Götterdämmerung", composer=composer)

    instance = Foo(
        bar=Bar(a="a", b=True),
        spam=Spam(thing=["one", "two", "three", "four"], again="ok"),
        slim_enum_type=SlimEnum,
        chunky_enum_type=ChunkyEnum,
        pick=composer,
        pick_many=list(models.Composer.objects.all()),
        pick_method=opera,
        pick_property=composer,
        pick_literal=composer,
        pick_computed_queryset=composer,
        pick_nested=composer,
        pick_enum=continent,
        fixed_tuple_different_types=("ok", 5),
        fixed_tuple_same_types=("alright", "again"),
        complex_type_we_do_not_type={
            "I am": "very complex",
            "because": {"I am": "nested", "ok?": []},
        },
    )
    definitions = {}
    generated_schema = create_schema(Foo, definitions)

    serialized = serialize(instance, generated_schema)

    assert serialized == {
        "bar": {"a": "a", "b": True},
        "spam": {"thing": ["one", "two", "three", "four"], "again": "ok"},
        "slim_enum_type": {"FIRST": "Ok", "SECOND": "Great"},
        "chunky_enum_type": {
            "CHUNKY_FIRST": {"is_important": False, "title": "Chunky First"},
            "CHUNKY_SECOND": {"is_important": True, "title": "Chunky Second"},
        },
        "pick": {"name": composer.name, "operas": [{"name": opera.name}]},
        "pick_many": [{"name": "Wagner", "operas": [{"name": "Götterdämmerung"}]}],
        "pick_method": {
            "name": "Götterdämmerung",
            "get_birthplace_of_composer": "Germany",
        },
        "pick_property": {"did_live_in_more_than_one_country": True,},
        "pick_literal": {"name": composer.name, "operas": [{"name": opera.name}]},
        "pick_computed_queryset": {
            "operas_with_piano_transcriptions": [{"name": "Götterdämmerung"}]
        },
        "pick_nested": {"name": composer.name, "operas": [{"name": opera.name}]},
        "pick_enum": {"hemisphere": "SOUTHERN"},
        "fixed_tuple_different_types": ["ok", 5],
        "fixed_tuple_same_types": ["alright", "again"],
        "complex_type_we_do_not_type": {
            "I am": "very complex",
            "because": {"I am": "nested", "ok?": []},
        },
    }

    convert_to_json_and_validate(serialized, generated_schema)


def test_form():
    generated_schema = create_schema(forms.OperaForm, {})
    form = forms.OperaForm()
    serialized_form = serialize(form, generated_schema)
    convert_to_json_and_validate(serialized_form, generated_schema)

    form_with_errors = forms.OperaForm({})
    form_with_errors.is_valid()
    serialized_form = serialize(form_with_errors, generated_schema)
    assert "name" in serialized_form.errors
    convert_to_json_and_validate(serialized_form, generated_schema)


def test_custom_widget():
    class CustomWidget(django_forms.Select):
        reactivated_widget = "foo"

    class CustomForm(django_forms.Form):
        field = django_forms.CharField(widget=CustomWidget)

    generated_schema = create_schema(CustomForm, {})
    form = CustomForm()
    serialized_form = serialize(form, generated_schema)
    convert_to_json_and_validate(serialized_form, generated_schema)
    assert serialized_form.fields["field"].widget["template_name"] == "foo"


@pytest.mark.django_db
def test_form_with_model_choice_iterator_value():
    models.Country.objects.create(
        name="USA",
        continent=models.Continent.objects.create(
            name="America", hemisphere="NORTHERN"
        ),
    )
    iterator = (
        forms.ComposerForm()
        .fields["countries"]
        .widget.optgroups("countries", "")[0][1][0]["value"]
    )

    assert isinstance(iterator, ModelChoiceIteratorValue)

    generated_schema = create_schema(forms.ComposerForm, {})
    form = forms.ComposerForm()
    serialized_form = serialize(form, generated_schema)
    convert_to_json_and_validate(serialized_form, generated_schema)
    assert serialized_form.fields["countries"].widget["optgroups"][0][1][0][
        "value"
    ] == str(iterator.value)


@pytest.mark.django_db
def test_form_set():
    generated_schema = create_schema(forms.OperaFormSet, {})
    form_set = forms.OperaFormSet()
    serialized_form_set = serialize(form_set, generated_schema)
    convert_to_json_and_validate(serialized_form_set, generated_schema)

    form_set_with_errors = forms.OperaFormSet(
        {"form-TOTAL_FORMS": 20, "form-INITIAL_FORMS": 0}
    )
    form_set_with_errors.is_valid()
    serialized_form_set = serialize(form_set_with_errors, generated_schema)
    assert "name" in serialized_form_set.forms[0].errors
    convert_to_json_and_validate(serialized_form_set, generated_schema)


def test_typed_choices_non_enum(settings):
    settings.INSTALLED_APPS = ["tests.serialization"]
    test_apps = Apps(settings.INSTALLED_APPS)

    class TestModel(Model):
        non_enum_typed_field = IntegerField(choices=((0, "Zero"), (1, "One")))

        class Meta:
            apps = test_apps

    class TestForm(django_forms.ModelForm):
        class Meta:
            model = TestModel
            fields = "__all__"

    generated_schema = create_schema(TestForm, {})
    assert generated_schema.definitions[
        "tests.serialization.test_typed_choices_non_enum.<locals>.TestForm"
    ]["properties"]["fields"]["properties"]["non_enum_typed_field"]["properties"][
        "widget"
    ] == {
        "tsType": "widgets.Select"
    }
