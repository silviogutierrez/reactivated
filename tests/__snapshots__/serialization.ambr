# serializer version: 1
# name: test_form_and_fields
  dict({
    'help_text': None,
    'label': 'Date field',
    'name': 'date_field',
    'widget': dict({
      'attrs': dict({
        'id': 'id_date_field',
        'required': True,
      }),
      'is_hidden': False,
      'name': 'date_field',
      'required': True,
      'subwidgets': list([
        dict({
          'attrs': dict({
            'id': 'id_date_field_month',
            'required': True,
          }),
          'is_hidden': False,
          'name': 'date_field_month',
          'optgroups': list([
            list([
              None,
              list([
                dict({
                  'label': 'January',
                  'name': 'date_field_month',
                  'selected': True,
                  'value': 1,
                }),
              ]),
              0,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'February',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 2,
                }),
              ]),
              1,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'March',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 3,
                }),
              ]),
              2,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'April',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 4,
                }),
              ]),
              3,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'May',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 5,
                }),
              ]),
              4,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'June',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 6,
                }),
              ]),
              5,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'July',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 7,
                }),
              ]),
              6,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'August',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 8,
                }),
              ]),
              7,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'September',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 9,
                }),
              ]),
              8,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'October',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 10,
                }),
              ]),
              9,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'November',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 11,
                }),
              ]),
              10,
            ]),
            list([
              None,
              list([
                dict({
                  'label': 'December',
                  'name': 'date_field_month',
                  'selected': False,
                  'value': 12,
                }),
              ]),
              11,
            ]),
          ]),
          'required': False,
          'tag': 'django.forms.widgets.Select',
          'template_name': 'django/forms/widgets/select.html',
          'value': '1',
        }),
        dict({
          'attrs': dict({
            'id': 'id_date_field_day',
            'required': True,
          }),
          'is_hidden': False,
          'name': 'date_field_day',
          'optgroups': list([
            list([
              None,
              list([
                dict({
                  'label': '1',
                  'name': 'date_field_day',
                  'selected': True,
                  'value': 1,
                }),
              ]),
              0,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '2',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 2,
                }),
              ]),
              1,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '3',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 3,
                }),
              ]),
              2,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '4',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 4,
                }),
              ]),
              3,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '5',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 5,
                }),
              ]),
              4,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '6',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 6,
                }),
              ]),
              5,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '7',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 7,
                }),
              ]),
              6,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '8',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 8,
                }),
              ]),
              7,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '9',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 9,
                }),
              ]),
              8,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '10',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 10,
                }),
              ]),
              9,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '11',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 11,
                }),
              ]),
              10,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '12',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 12,
                }),
              ]),
              11,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '13',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 13,
                }),
              ]),
              12,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '14',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 14,
                }),
              ]),
              13,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '15',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 15,
                }),
              ]),
              14,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '16',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 16,
                }),
              ]),
              15,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '17',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 17,
                }),
              ]),
              16,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '18',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 18,
                }),
              ]),
              17,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '19',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 19,
                }),
              ]),
              18,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '20',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 20,
                }),
              ]),
              19,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '21',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 21,
                }),
              ]),
              20,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '22',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 22,
                }),
              ]),
              21,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '23',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 23,
                }),
              ]),
              22,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '24',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 24,
                }),
              ]),
              23,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '25',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 25,
                }),
              ]),
              24,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '26',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 26,
                }),
              ]),
              25,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '27',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 27,
                }),
              ]),
              26,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '28',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 28,
                }),
              ]),
              27,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '29',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 29,
                }),
              ]),
              28,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '30',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 30,
                }),
              ]),
              29,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '31',
                  'name': 'date_field_day',
                  'selected': False,
                  'value': 31,
                }),
              ]),
              30,
            ]),
          ]),
          'required': False,
          'tag': 'django.forms.widgets.Select',
          'template_name': 'django/forms/widgets/select.html',
          'value': '1',
        }),
        dict({
          'attrs': dict({
            'id': 'id_date_field_year',
            'required': True,
          }),
          'is_hidden': False,
          'name': 'date_field_year',
          'optgroups': list([
            list([
              None,
              list([
                dict({
                  'label': '2000',
                  'name': 'date_field_year',
                  'selected': False,
                  'value': 2000,
                }),
              ]),
              0,
            ]),
            list([
              None,
              list([
                dict({
                  'label': '2001',
                  'name': 'date_field_year',
                  'selected': False,
                  'value': 2001,
                }),
              ]),
              1,
            ]),
          ]),
          'required': False,
          'tag': 'django.forms.widgets.Select',
          'template_name': 'django/forms/widgets/select.html',
          'value': '2015',
        }),
      ]),
      'tag': 'django.forms.widgets.SelectDateWidget',
      'template_name': 'django/forms/widgets/select_date.html',
      'value': dict({
        'day': 1,
        'month': 1,
        'year': 2015,
      }),
    }),
  })
# ---
# name: test_form_and_fields.1
  dict({
    'help_text': None,
    'label': 'Date time field',
    'name': 'date_time_field',
    'widget': dict({
      'attrs': dict({
        'id': 'id_date_time_field',
        'required': True,
      }),
      'is_hidden': False,
      'name': 'date_time_field',
      'required': True,
      'subwidgets': list([
        dict({
          'attrs': dict({
            'id': 'id_date_time_field_0',
            'required': True,
          }),
          'is_hidden': False,
          'name': 'date_time_field_0',
          'required': False,
          'tag': 'django.forms.widgets.DateInput',
          'template_name': 'django/forms/widgets/date.html',
          'type': 'text',
          'value': None,
        }),
        dict({
          'attrs': dict({
            'id': 'id_date_time_field_1',
            'required': True,
          }),
          'is_hidden': False,
          'name': 'date_time_field_1',
          'required': False,
          'tag': 'django.forms.widgets.TimeInput',
          'template_name': 'django/forms/widgets/time.html',
          'type': 'text',
          'value': None,
        }),
      ]),
      'tag': 'django.forms.widgets.SplitDateTimeWidget',
      'template_name': 'django/forms/widgets/splitdatetime.html',
      'value': None,
    }),
  })
# ---
# name: test_form_and_fields.2
  dict({
    'help_text': 'Not blank',
    'label': 'Boolean field',
    'name': 'boolean_field',
    'widget': dict({
      'attrs': dict({
        'checked': True,
        'id': 'id_boolean_field',
        'required': True,
      }),
      'is_hidden': False,
      'name': 'boolean_field',
      'required': True,
      'tag': 'django.forms.widgets.CheckboxInput',
      'template_name': 'django/forms/widgets/checkbox.html',
      'type': 'checkbox',
      'value': True,
    }),
  })
# ---
# name: test_nested_null_foreign_keys
  dict({
    'test_nested_null_foreign_keys.<locals>.Child_cf6d15b34c': dict({
      'additionalProperties': False,
      'properties': dict({
        'parent': dict({
          'anyOf': list([
            dict({
              'additionalProperties': False,
              'properties': dict({
                'grandparent': dict({
                  'anyOf': list([
                    dict({
                      'additionalProperties': False,
                      'properties': dict({
                        'description': dict({
                          'type': 'string',
                        }),
                        'id': dict({
                          'type': 'number',
                        }),
                        'name': dict({
                          'type': 'string',
                        }),
                      }),
                      'required': list([
                        'id',
                        'name',
                        'description',
                      ]),
                      'type': 'object',
                    }),
                    dict({
                      'type': 'null',
                    }),
                  ]),
                }),
              }),
              'required': list([
                'grandparent',
              ]),
              'type': 'object',
            }),
            dict({
              'type': 'null',
            }),
          ]),
        }),
      }),
      'required': list([
        'parent',
      ]),
      'title': 'test_nested_null_foreign_keys.<locals>.Child_cf6d15b34c',
      'type': 'object',
    }),
  })
# ---
# name: test_null_one_to_one_field
  dict({
    'test_null_one_to_one_field.<locals>.Brother_56906733d0': dict({
      'additionalProperties': False,
      'properties': dict({
        'sister': dict({
          'anyOf': list([
            dict({
              'additionalProperties': False,
              'properties': dict({
                'id': dict({
                  'type': 'number',
                }),
              }),
              'required': list([
                'id',
              ]),
              'type': 'object',
            }),
            dict({
              'type': 'null',
            }),
          ]),
        }),
      }),
      'required': list([
        'sister',
      ]),
      'title': 'test_null_one_to_one_field.<locals>.Brother_56906733d0',
      'type': 'object',
    }),
  })
# ---
# name: test_override_pick_types
  dict({
    'additionalProperties': False,
    'properties': dict({
      'forced_non_nullable': dict({
        'type': 'number',
      }),
      'forced_none': dict({
        'type': 'null',
      }),
      'forced_nullable': dict({
        '_reactivated_union': dict({
          'builtins.int': dict({
            'type': 'number',
          }),
        }),
        'anyOf': list([
          dict({
            'type': 'number',
          }),
          dict({
            'type': 'null',
          }),
        ]),
        'serializer': 'reactivated.serialization.UnionType',
      }),
    }),
    'required': list([
      'forced_nullable',
      'forced_non_nullable',
      'forced_none',
    ]),
    'title': 'test_override_pick_types.<locals>.TestModel_fdae93dc88',
    'type': 'object',
  })
# ---
# name: test_serialize_lazy_object
  Thing(
    definitions=dict({
    }),
    schema=dict({
      'type': 'string',
    }),
  )
# ---
# name: test_simple_union
  Thing(
    definitions=dict({
    }),
    schema=dict({
      '_reactivated_union': dict({
        'builtins.bool': dict({
          'type': 'boolean',
        }),
        'builtins.int': dict({
          'type': 'number',
        }),
        'builtins.list': dict({
          'items': dict({
            'type': 'string',
          }),
          'type': 'array',
        }),
        'builtins.str': dict({
          'type': 'string',
        }),
        'builtins.tuple': dict({
          'items': list([
            dict({
              'type': 'number',
            }),
            dict({
              'type': 'string',
            }),
          ]),
          'maxItems': 2,
          'minItems': 2,
          'type': 'array',
        }),
        'datetime.date': dict({
          'type': 'string',
        }),
      }),
      'anyOf': list([
        dict({
          'type': 'string',
        }),
        dict({
          'items': dict({
            'type': 'string',
          }),
          'type': 'array',
        }),
        dict({
          'items': list([
            dict({
              'type': 'number',
            }),
            dict({
              'type': 'string',
            }),
          ]),
          'maxItems': 2,
          'minItems': 2,
          'type': 'array',
        }),
        dict({
          'type': 'number',
        }),
        dict({
          'type': 'string',
        }),
        dict({
          'type': 'boolean',
        }),
        dict({
          'type': 'null',
        }),
      ]),
      'serializer': 'reactivated.serialization.UnionType',
    }),
  )
# ---
# name: test_union_with_literal
  Thing(
    definitions=dict({
    }),
    schema=dict({
      '_reactivated_union': dict({
        'literal-0-0.builtins.str': dict({
          'enum': list([
            'one',
          ]),
        }),
        'literal-0-1.builtins.str': dict({
          'enum': list([
            'three',
          ]),
        }),
        'literal-0-2.builtins.int': dict({
          'enum': list([
            3,
          ]),
        }),
        'literal-0-3.builtins.bool': dict({
          'enum': list([
            False,
          ]),
        }),
      }),
      'anyOf': list([
        dict({
          'enum': list([
            'one',
          ]),
        }),
        dict({
          'enum': list([
            'three',
          ]),
        }),
        dict({
          'enum': list([
            3,
          ]),
        }),
        dict({
          'enum': list([
            False,
          ]),
        }),
        dict({
          'type': 'null',
        }),
      ]),
      'serializer': 'reactivated.serialization.UnionType',
    }),
  )
# ---
# name: test_uuid_field
  dict({
    'serializer': 'reactivated.serialization.UUIDFieldType',
    'tsType': 'UUID',
  })
# ---
# name: test_uuid_field.1
  dict({
  })
# ---
# name: test_uuid_field.2
  dict({
    'anyOf': list([
      dict({
        'serializer': 'reactivated.serialization.UUIDFieldType',
        'tsType': 'UUID',
      }),
      dict({
        'type': 'null',
      }),
    ]),
  })
# ---
# name: test_uuid_field.3
  dict({
  })
# ---
